text/0201-scope-attributes.md

```diff
@@ -0,0 +1,148 @@
+# Introduce Scope Attributes
+
+This OTEP adds attributes to the Scope.
+
+## Motivation
+
+There are a few reasons why adding Scope attributes is a good idea:
+
+- There are 2 known use cases where Scope attributes can solve specific problems:
+  - Add support for [Meter "short_name"](https://github.com/open-telemetry/opentelemetry-specification/pull/2422),
+    represented as an attribute of Meter's Scope.
+  - Add support for differentiating the type of data emitted from the scopes that belong
+    to different data domains, e.g. profiling data emitted as log records or client-side
+    data emitted as log records needs to be differentiated so that it can be easily
+    routed and processed differently in the backends. We don't have a good way to handle
+    this today. The type of the data can be recorded as an attribute Logger's Scope.
+- It makes Scope consistent with the other primary data types: Resource, Span, Metric,
+  LogRecord.
+
+See additional [discussion here](https://github.com/open-telemetry/opentelemetry-specification/issues/2450).
+
+## Summary
+
+The following is the summary of proposed changes:
+
+- We will extend OpenTelemetry API to allow specifying Scope attributes when obtaining a
+  Tracer, Meter or LogEmitter. Scope attributes will be optional.
+- We will add `attributes` field to the [InstrumentationScope](https://github.com/open-telemetry/opentelemetry-proto/blob/88faab1197a2a105c7da659951e94bc951d37ab9/opentelemetry/proto/common/v1/common.proto#L83)
+  message of OTLP.
+- We will specify that Telemetry emitted via a Scope-ed Tracer, Meter or LogEmitter will
+  be associated with the Scope's attributes.
+- We will specify that OTLP Exporter will record the attributes in the
+  InstrumentationScope message.
+- We will create a section for Scope attributes' semantic conventions in
+  the specification.
+
+## Internal details
+
+### API Changes
+
+`Get a Tracer` API will be extended to add the following parameter:
+
+```
+- `attributes` (optional): Specifies the instrumentation scope attributes to associate
+  with emitted telemetry.
+```
+
+`Get a Metter` API will be extended to add the following parameter:
+
+```
+- `attributes` (optional): Specifies the instrumentation scope attributes to associate
+  with emitted telemetry.
```

Comment:
My intuition was that it would be cheaper to specify the scope attribute once when obtaining the Tracer/Meter/LogEmitter (I may be wrong). The other benefit is that for the use cases I had in my mind I would want to specify the scope attributes once.

You are right though that there may be use cases where scope attributes may need to be specifying for every data point. An example of these may be the client-side session id that was discussed in the SIG meeting today. I think it largely depends on how exactly the client-side SIG intends to perform the instrumentation, where the session id is available and what is the most ergonomic way to set it on the scope.

I think we need feedback from multiple language maintainers to help shape this particular API. It may be that we need more than one way to specify the scope attributes (however, I would prefer to have only one if possible).

Reply:
> You are right though that there may be use cases where scope attributes may need to be specifying for every data point. An example of these may be the client-side session id that was discussed in the SIG meeting today

I was under the impression that session id is global (at a specific point in time) to the application and affects all the telemetry data collected, no matter what scope it is coming from. Thus it made sense to me to include it in the resource.
I wonder when would someone want to store a session id on one scope and omit it in another.

Reply:
There is a separate discussion about session id elsewhere: https://docs.google.com/document/d/1bsS8uAXv1U-i97dj5PP7LtVxWp08mmYXlah6XDVS-xQ/edit
It may or may not be solved by this new Scope attributes.

Reply:
> I think we need feedback from multiple language maintainers to help shape this particular API. It may be that we need more than one way to specify the scope attributes (however, I would prefer to have only one if possible).

Agreed!